### ISSUE TITLE
valid-metrics: bleu-detok issue (with sentencepiece)

### ISSUE THREAD

**Issue by geovedi** [marian-nmt/marian#233]
currently i'm in the process of retraining my old dataset (en-id) with base transformer configuration and sentencepiece. i noticed that cross-entropy and perplexity validation results are improved but not blue-detok. i quickly double checked by evaluating the translation result with sacrebleu, the training was actually improved.

the following are the snippet from the train.log and manual check with sacrebleu:
```
[2018-12-12 18:04:33] Ep. 1 : Up. 18000 : Sen. 13,161,686 : Cost 2.09789205 : Time 1277.30s : 13729.39 words/s : L.r. 2.8284e-04
[2018-12-12 18:25:54] Ep. 1 : Up. 19000 : Sen. 13,885,795 : Cost 2.07635450 : Time 1280.97s : 13697.95 words/s : L.r. 2.7530e-04
[2018-12-12 18:47:21] Ep. 1 : Up. 20000 : Sen. 14,615,612 : Cost 2.05718994 : Time 1287.42s : 13695.46 words/s : L.r. 2.6833e-04
[2018-12-12 18:47:36] [valid] Ep. 1 : Up. 20000 : cross-entropy : 48.3253 : new best
[2018-12-12 18:47:43] [valid] Ep. 1 : Up. 20000 : perplexity : 24.316 : new best
[2018-12-12 18:48:48] [valid] Ep. 1 : Up. 20000 : bleu-detok : 7.9485 : new best
```

```
$ ./sacreBLEU/sacrebleu.py data/valid.id < data/valid.en.id
BLEU+case.mixed+numrefs.1+smooth.exp+tok.13a+version.1.2.12 = 58.8 83.8/66.7/53.9/43.9 (BP = 0.975 ratio = 0.975 hyp_len = 117299 ref_len = 120247)
```

```
[2018-12-12 21:39:33] Ep. 1 : Up. 28000 : Sen. 20,467,137 : Cost 1.96469533 : Time 1277.83s : 13685.25 words/s : L.r. 2.2678e-04
[2018-12-12 22:00:47] Ep. 1 : Up. 29000 : Sen. 21,195,575 : Cost 1.95809531 : Time 1274.33s : 13700.93 words/s : L.r. 2.2283e-04
[2018-12-12 22:22:05] Ep. 1 : Up. 30000 : Sen. 21,930,504 : Cost 1.95560884 : Time 1278.32s : 13724.65 words/s : L.r. 2.1909e-04
[2018-12-12 22:22:18] [valid] Ep. 1 : Up. 30000 : cross-entropy : 40.0537 : new best
[2018-12-12 22:22:23] [valid] Ep. 1 : Up. 30000 : perplexity : 14.0824 : new best
[2018-12-12 22:23:26] [valid] Ep. 1 : Up. 30000 : bleu-detok : 7.86407 : stalled 1 times
```

```
$ ./sacreBLEU/sacrebleu.py data/valid.id < data/valid.en.id
BLEU+case.mixed+numrefs.1+smooth.exp+tok.13a+version.1.2.12 = 59.6 84.1/67.3/54.6/44.7 (BP = 0.977 ratio = 0.977 hyp_len = 117510 ref_len = 120247)
```

both marian and marian-dev (all masters) are showing the same behaviour.

is this expected?

**Comment by emjotde** [comment]
Hm, no. Also the numbers should be roughly the same. No idea what's wrong though. Can you post the whole log?

**Comment by geovedi** [comment]
sure...

[train.log](https://github.com/marian-nmt/marian/files/2672681/train.log)
[valid.log](https://github.com/marian-nmt/marian/files/2672682/valid.log)

**Comment by emjotde** [comment]
There we go:
```
[2018-12-12 15:02:58] [config] valid-sets:
[2018-12-12 15:02:58] [config]   - data/valid.en
[2018-12-12 15:02:58] [config]   - data/valid.en
```
Same valid set on both sides.

**Comment by geovedi** [comment]
duh!

**Comment by emjotde** [comment]
@snukky Should we add a warning at each validation step when the files are the same?

**Comment by snukky** [comment]
I just wanted to ask this question :) So, yes, I will add this.

**Comment by geovedi** [comment]
shouldn't sanity check done at early stage?

**Comment by emjotde** [comment]
Both, I guess. We should still keep this as a warning not an abort as there is a number of situations, where this might actually make sense.

**Comment by wingsyuan** [comment]
@emjotde 
in Tutorial: Marian with Built-in SentencePiece, you mentioned that you will  soon update WMT Transformer examples to use SentencePiece, looking forward..
Can you help me checking whether the following command will work well with Transformer using sentencepiece ?  simply add the option "--sentencepiece-alphas 0.2 0 \"
Thanks you so much!

 $MARIAN/marian \
    --type transformer \
    --model model-sentenpiece/model.npz \
    --train-sets ../clean_uniq_train.zh ../clean_uniq_train.en \
    --max-length 200 \
    --vocabs model-sentenpiece/vocab.zhen.spm model-sentenpiece/vocab.zhen.spm \
    --sentencepiece-alphas 0.2 0 \
    --dim-vocabs 32000 32000 \
    --mini-batch-fit -w $WORKSPACE \
    --valid-freq 2000 --save-freq 2000  --disp-freq 500 \
    --valid-metrics ce-mean-words perplexity translation \
    --valid-sets ../valid_laws_1000.zh ../valid_laws_1000.en \
    --valid-script-path "bash ./scripts/validate.sh" \
    --valid-translation-output vali.output --quiet-translation \
    --beam-size 5 --normalize=1 \
    --valid-mini-batch 32 \
    --overwrite --keep-best \
    --early-stopping 5 --after-epochs $EPOCHS --cost-type=ce-mean-words \
    --log train.log --valid-log valid.log \
    --enc-depth 6 --dec-depth 6 \
    --tied-embeddings-all \
    --transformer-dropout 0.1 --label-smoothing 0.1 \
    --learn-rate 0.0003 --lr-warmup 16000 --lr-decay-inv-sqrt 16000 --lr-report \
    --optimizer-params 0.9 0.98 1e-09 --clip-norm 5 \
    --devices $GPUS --sync-sgd --seed 1111  \
    --exponential-smoothing

**Comment by jgochal** [comment]
To revisit wingsyuan's question, has anyone been able to get a Transformer model working properly with Sentencepiece?  I've tried the same and similar setups and the training basically stalls quite early on in epoch 1 with a valid BLEU score of 0.
